katello-attach-subscription

```diff
@@ -99,70 +139,659 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+          next
+        end
+        # if no guests present, retrieve all subscription data from that hypervisor
+        begin
+          rem_sub_response = @api.resource(:host_subscriptions).call(:index, {:host_id => system['id']})
+          if @options[:debug]
+            puts "   DEBUG: Retrieved subscription data of #{system['name']}"
+            p rem_sub_response
+          end
+        rescue RestClient::NotFound
+          STDERR.puts "   ERROR: Hypervisor #{system['name']} subscription data not found. Skipping."
+          next
+        rescue Exception => e
+          puts "   ERROR: Unknow Error -- Unable to retrieve subscription data for hypervisor #{system['id']} #{system['name']}"
+          puts e.message
+          puts e.backtrace.inspect
+          puts e.response
+          exit 1
+        end
+
+        # check if data found (API return data successfully)
+        if not rem_sub_response.has_key?("total")
+          if @options[:verbose]
+            puts "    VERBOSE: Skipping Hypervisor as not valid data was retrieved from API"
+          end
+          next
+        end
+
+        # check if total > 0, then loop the result to remove the subscriptions
+        if rem_sub_response["total"].to_i > 0
+          puts "    Starting removing subscriptions from Hypervisor"
+          rem_sub_response["results"].each_with_index do |rem_sub, currentsub|
+            if @options[:verbose]
+              puts "      VERBOSE: Removing subscription #{currentsub+1} / #{rem_sub_response["total"].to_i} - #{rem_sub['id']}"
+            end
+            if @options[:debug]
+              puts "      DEBUG: Current subscription data"
+              p rem_sub
+            end
+            begin
+              # removing the subscription calling remove_subscriptions API
+              if not @options[:noop]
+                @api.resource(:host_subscriptions).call(:remove_subscriptions, {:host_id => system['id'], :subscriptions => [{:id => rem_sub['id']}]})
+                puts "      Removed subscription #{rem_sub['id']} from hypervisor #{system['name']} / #{system['id']}"
+              else
+                puts "      [noop]: sub #{rem_sub['id']} would be removed from hypervisor #{system['name']} / #{system['id']}"
+              end
+            rescue Exception => e
+              puts "      ERROR: Unknow Error - Unable to remove subscription #{rem_sub["id"]} from #{system["name"]}"
+              puts e.message
+              puts e.backtrace.inspect
+              puts e.response
+              exit 1
+            end
+          end
+        else
+          puts "    No subscription found attached to #{system['name']}. Nothing to do."
+        end
+      end
+    else
+      if @options[:verbose]
+        puts "  VERBOSE: No empty hypervisor with valid subscription found."
+      end
+    end
+  end
+
+end
+
+
+# main function
+def vdcupdate()
+  # satellite url has to start with https or PUT will fail with http error
```

Comment:
I wonder if this should move outside of `vdcupdate` now that we have two functions we call.